pkg/vpcmodel/semanticDiffSubnets.go

```diff
@@ -0,0 +1,212 @@
+package vpcmodel
+
+import (
+	"fmt"
+
+	"github.com/np-guard/vpc-network-config-analyzer/pkg/common"
+)
+
+// ToDo: go over structs specifically * and lack of
+
+type DiffType = int
+
+const (
+	NoDiff DiffType = iota
+	MissingSrcEP
+	MissingDstEP
+	MissingSrcDstEP
+	MissingConnection
+	ChangedConnection
+)
+
+type connectionDiff struct {
+	*common.ConnectionSet
+	diff DiffType
+}
+
+type SubnetsDiff map[EndpointElem]map[EndpointElem]*connectionDiff
+
+type ConfigsForDiff struct {
+	config1 *CloudConfig
+	config2 *CloudConfig
+}
+
+type SubnetConfigConnectivity struct {
+	config             *CloudConfig
+	subnetConnectivity SubnetConnectivityMap
+}
+
+type diffBetweenSubnets struct {
+	subnet1Subtract2 SubnetsDiff
+	subnet2Subtract1 SubnetsDiff
+
+	GroupedSubnet1Minus2 *GroupConnLines
+	GroupedSubnet1Minus1 *GroupConnLines
+}
+
+func (configs ConfigsForDiff) GetSubnetsDiff(grouping bool) (*diffBetweenSubnets, error) {
+	// 1. compute connectivity for each of the subnets
+	subnetsConn1, err := configs.config1.GetSubnetsConnectivity(true, grouping)
+	if err != nil {
+		return nil, nil
+	}
+	subnetsConn2, err := configs.config2.GetSubnetsConnectivity(true, grouping)
+	if err != nil {
+		return nil, nil
+	}
+
+	// 2. Computes delta in both directions
+	subnet1Aligned, subnet2Aligned := subnetsConn1.AllowedConnsCombined.getConnectivesWithSameIPBlocks(subnetsConn2.AllowedConnsCombined)
+	subnetConfigConnectivity1 := SubnetConfigConnectivity{configs.config1, subnet1Aligned}
+	subnetConfigConnectivity2 := SubnetConfigConnectivity{configs.config2, subnet2Aligned}
+	subnet1Subtract2 := subnetConfigConnectivity1.SubnetConnectivitySubtract(&subnetConfigConnectivity2)
+	subnet2Subtract1 := subnetConfigConnectivity2.SubnetConnectivitySubtract(&subnetConfigConnectivity1)
+
+	// 3. ToDo: grouping, see comment at the end of this file
+
+	res := &diffBetweenSubnets{
+		subnet1Subtract2: subnet1Subtract2,
+		subnet2Subtract1: subnet2Subtract1}
+	return res, nil
+}
+
+// for a given EndpointElem (representing a subnet or an external ip) in config return the EndpointElem representing the
+// subnet/external address in otherConfig or nil if the subnet does not exist in the other config.
+// ToDo: this is done based on names only at the moment. Perhaps take into account other factors such as cidr?
+// ToDo: instead of performing this search each time, use a map created once
+func (c *CloudConfig) getEndpointElemInOtherConfig(other *CloudConfig, ep EndpointElem) EndpointElem {
+	if ep.IsExternal() {
+		for _, node := range other.Nodes {
+			if node.Name() == ep.Name() {
+				res := EndpointElem(node)
+				return res
+			}
+		}
```

Comment:
I don't think that external ip-blocks can be considered as "missing endpoint" in a config. The connectivity result per config should contain connectivity per all ip-blocks ranges.

Reply:
The current PR and its unit test does not actually handle external ranges. So let me comment it for now and add todo, and will be handled in the next PR that handles external addresses.

Reply:
fixed